e2etest/localserviceimpact_test.go

```diff
@@ -0,0 +1,280 @@
+		var slices *discoveryv1.EndpointSliceList
+		Eventually(func() int {
+			eps := 0
+			slices, err = cluster2.k8s.DiscoveryV1().EndpointSlices(namespace).List(ctx, metav1.ListOptions{
+				LabelSelector: labels.Set{discoveryv1.LabelServiceName: helloService.Name}.AsSelector().String(),
+			})
+			Expect(err).ToNot(HaveOccurred())
+			for _, s := range slices.Items {
+				eps += len(s.Endpoints)
+			}
+			return eps
+		}, 30).Should(Equal(1))
+		importedSlice := slices.Items[0]
+		importedSlice.ObjectMeta = metav1.ObjectMeta{
+			Name: importedSlice.Name,
+			Labels: map[string]string{
+				v1alpha1.LabelServiceName: helloServiceImport.Name,
+			},
+		}
+		_, err = cluster1.k8s.DiscoveryV1().EndpointSlices(namespace).Create(ctx, &importedSlice, metav1.CreateOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		Eventually(func() []string {
+			svcImport, err := cluster1.mcs.MulticlusterV1alpha1().ServiceImports(namespace).Get(ctx, helloServiceImport.Name, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			return svcImport.Spec.IPs
+		}).ShouldNot(BeEmpty())
+		Eventually(func() string {
+			updatedSlice, err := cluster1.k8s.DiscoveryV1().EndpointSlices(namespace).Get(ctx, importedSlice.Name, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			return updatedSlice.Labels[discoveryv1.LabelServiceName]
+		}).ShouldNot(BeEmpty())
+		reqPod, err = cluster1.k8s.CoreV1().Pods(namespace).Get(ctx, requestPod.Name, metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		By("Created all in " + namespace)
+	})
+	AfterEach(func() {
+		if *noTearDown {
+			By(fmt.Sprintf("Skipping tearndown. Test namespace %q", namespace))
+			By(fmt.Sprintf("Cluster 1: kubectl --kubeconfig %q -n %q", *kubeconfig1, namespace))
+			By(fmt.Sprintf("Cluster 2: kubectl --kubeconfig %q -n %q", *kubeconfig2, namespace))
+			return
+		}
+		Expect(cluster1.k8s.CoreV1().Namespaces().Delete(ctx, namespace, metav1.DeleteOptions{}))
+		Expect(cluster2.k8s.CoreV1().Namespaces().Delete(ctx, namespace, metav1.DeleteOptions{}))
+	})
+	Specify("DNS resolves as expected", func() {
+		By("verifying a local, unexported service is reachable via cluster.local")
+		successfulChecks := 0
+		command := []string{"sh", "-c", fmt.Sprintf("echo hi | nc %s.%s.svc.cluster.local 42", helloService.Name, namespace)}
+		for i := 0; i <= 59; i++ {
+			stout, _, err := execCmd(cluster1.k8s, restcfg1, reqPod.Name, reqPod.Namespace, command)
+			Expect(err).ToNot(HaveOccurred())
+			clusterID := strings.TrimSpace(string(stout))
+			if clusterID == strings.TrimSpace("cluster1") {
+				successfulChecks++
+			}
+		}
+		Eventually(func() int {
+			return successfulChecks
+		}).Should(BeNumerically(">=", 50))
+
+		By("verifying a remote, exported service is reachable via clusterset.local")
+		successfulChecks = 0
+		command = []string{"sh", "-c", fmt.Sprintf("echo hi | nc %s.%s.svc.clusterset.local 42", helloServiceImport.Name, namespace)}
+		for i := 0; i <= 59; i++ {
```

Comment:
This feels like it could be more idiomatically written as a combination of `Eventually` and `MustPassRepeatedly`: first the check must succeed (eventually), then it must pass repeatedly.

The same applies to the similar construct above. The core check should be extracted to a separate function.